python/giotto_json_reader.py

```diff
@@ -0,0 +1,134 @@
+#!/usr/bin/env python3
+
+import json
+import argparse
+
+
+def cells_json(data):
+    '''
+    >>> data = {
+    ...     "cell_1": {
+    ...         "locations": [1632.02, -1305.7],
+    ...         "factors": {
+    ...             "pleiden_clus": [3],
+    ...             "kmeans": [8]
+    ...         },
+    ...         "mappings": {
+    ...             "tsne": [-11.0776, 6.0311],
+    ...             "umap": [-6.858, 15.7691]
+    ...         }
+    ...     }
+    ... }
+    >>> cells = cells_json(data)
+    >>> cells.keys()
+    dict_keys(['cell_1'])
+    >>> cells['cell_1'].keys()
+    dict_keys(['mappings', 'genes', 'xy', 'factors', 'poly'])
+
+    '''
+    cell_dict = {}
+    for cell in data.keys():
+
+        mappings_dict = {
+            't-SNE': data[cell]['mappings']['tsne'],
+            'UMAP': data[cell]['mappings']['umap']
+            }
+
+        genes_dict = {}
+
+        factors_dict = {
+            'pleiden_clus': 'Cluster {}'.format(
+                data[cell]['factors']['pleiden_clus'][0]
+            ),
+            'kmeans': 'Cluster {}'.format(data[cell]['factors']['kmeans'][0])
+            }
+
+        cell_dict[cell] = {
+            'mappings': mappings_dict,
+            'genes': genes_dict,
+            'xy': data[cell]['locations'],
+            'factors': factors_dict,
+            'poly': []
+        }
+
+    return cell_dict
+
+
+def factors_json(data):
+    '''
+    >>> data = {
+    ...     "cell_1": {
+    ...         "locations": [1632.02, -1305.7],
+    ...         "factors": {
+    ...             "pleiden_clus": [3],
+    ...             "kmeans": [8]
+    ...         },
+    ...         "mappings": {
+    ...             "tsne": [-11.0776, 6.0311],
+    ...             "umap": [-6.858, 15.7691]
+    ...         }
+    ...     }
+    ... }
+    >>> factors = factors_json(data)
+    >>> factors.keys()
+    dict_keys(['pleiden_clus', 'kmeans'])
+    >>> factors['kmeans'].keys()
+    dict_keys(['map', 'cells'])
+
+    '''
+    pleiden_clusters = set()
+    kmeans_clusters = set()
+    for cell in data:
+        factors = data[cell]['factors']
+        pleiden_clusters.add(factors['pleiden_clus'][0])
+        kmeans_clusters.add(factors['kmeans'][0])
+
+    pleiden_clusters = list(pleiden_clusters)
+    kmeans_clusters = list(kmeans_clusters)
```

Comment:
Reassignment is generally a bad idea: When the meaning of a variable changes across the code, it can be harder to understand what it means on any particular line. Exceptions might be incrementing or appending, but even then there are probably better approaches. `pleiden_clusters_set` and `pleiden_clusters_list` might be better.